### PR TITLE
Unify authentication screens

### PR DESCRIPTION
**Pull request by nmattia**
This unifies both authentication flows (to II and to other dapps) by showing the same authentication screen in both cases (with slight variations). The "welcome" page is also simplified to lead new users towards the correct flow, i.e. creating an anchor, while hiding everything else (unnecessary stuff for new users like anchor recovery, etc) behind another button. The welcome screen is only shown to new users.

Note: this also cleans up and restructures some of the authentication code but we'll need another cleaning pass for module and test structure. Will do separately.

Builds on top of https://github.com/dfinity/internet-identity/pull/993, https://github.com/dfinity/internet-identity/pull/996, https://github.com/dfinity/internet-identity/pull/999, https://github.com/dfinity/internet-identity/pull/1005 and https://github.com/dfinity/internet-identity/pull/1006.

### CODE REVIEW

File: src/frontend/src/flows/welcome.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import { setUserNumber } from "../utils/userNumber";
+import { unreachable } from "../utils/utils";
+import { Connection } from "../utils/iiConnection";
+import { LoginFlowResult, LoginData } from "../flows/login/flowResult";
+import { displayError } from "../components/displayError";
+import { footer } from "../components/footer";
+import { icLogo } from "../components/icons";
+import { auth } from "../flows/login";
+
+/** Properties of the "Welcome" screen, which welcome new users. */
```
Comment (nmattia) [inline src/frontend/src/flows/welcome.ts:12]: typo

**Comment by peterpeterparker** [comment]
There is a lack of space in `register.png` between title and sentence (according screenshots)

File: src/frontend/src/components/authenticateBox.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+import { displayError } from "./displayError";
+import { useRecovery } from "../flows/recovery/useRecovery";
+import { registerIfAllowed } from "../utils/registerAllowedCheck";
+import { withRef } from "../utils/lit-html";
+
+/** Properties for the authentication box component which authenticates a user
+ * to II or to another dapp */
+type PageProps = { templates: AuthTemplates } & {
+  onContinue: (arg: bigint) => void;
+  recoverAnchor: (userNumner?: bigint) => void;
```
Comment (peterpeterparker) [inline src/frontend/src/components/authenticateBox.ts:27]: typo: `userNumner` jawohl

File: src/frontend/src/components/authenticateBox.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+} from "../flows/login/flowResult";
+import { displayError } from "./displayError";
+import { useRecovery } from "../flows/recovery/useRecovery";
+import { registerIfAllowed } from "../utils/registerAllowedCheck";
+import { withRef } from "../utils/lit-html";
+
+/** Properties for the authentication box component which authenticates a user
+ * to II or to another dapp */
+type PageProps = { templates: AuthTemplates } & {
+  onContinue: (arg: bigint) => void;
```
Comment (peterpeterparker) [inline src/frontend/src/components/authenticateBox.ts:26]: should `arg` be typed or be more explicit? just feel a bit generic

File: src/frontend/src/components/authenticateBox.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+export const authenticate = async (
+  connection: Connection,
+  userNumber: bigint
+): Promise<LoginFlowSuccess | LoginFlowError> => {
+  try {
+    const result = await withLoader(() => connection.login(userNumber));
+    return apiResultToLoginFlowResult(result);
+  } catch (error) {
+    return {
```
Comment (peterpeterparker) [inline src/frontend/src/components/authenticateBox.ts:147]: not in this pr but doesn't this comes often in the code base? an utils can maybe be extracted to map error

File: src/frontend/src/components/authenticateBox.ts [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+  /* the (purely visual) arrow on the chasm */
+  const chasmToggleRef: Ref<HTMLSpanElement> = createRef();
+
+  /* Toggle the chasm open/closed */
+  const chasmToggle = () =>
+    withRef(chasmRef, (chasm) => {
+      const classes = chasm.classList;
+
+      if (classes.contains("c-chasm--closed")) {
```
Comment (peterpeterparker) [inline src/frontend/src/components/authenticateBox.ts:168]: not in this pr but would be neat to have a utils/helper to toggle classes I guess

File: src/frontend/src/flows/authenticate/index.ts [hunk 6]
```diff
@@ -1,229 +1,38 @@
...
+    <span class="t-strong">${origin}</span><br />
+  </p>`;
+  const chasm =
+    derivationOrigin !== undefined
+      ? {
+          info: "shared identity",
+          message: html`<span class="t-strong">${origin}</span> is an
+            alternative domain of <br /><span class="t-strong"
+              >${derivationOrigin}</span
+            ><br />and you will be authenticated to both with the same identity. `,
```
Comment (peterpeterparker) [inline src/frontend/src/flows/authenticate/index.ts:31]: the space after `identity. ` is needed? (most important question of the day I know...)

File: src/frontend/src/flows/welcome.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  register,
+  signin,
+}: WelcomeProps): TemplateResult => html` <section
+    class="l-container c-card c-card--highlight"
+    aria-label="Authentication"
+  >
+    <div class="c-logo">${icLogo}</div>
+    <p class="t-paragraph t-centered">Unlock the full power of Web3</p>
+    <article class="l-stack">
+      <div class="l-stack">
```
Comment (peterpeterparker) [inline src/frontend/src/flows/welcome.ts:40]: you need this div?

File: src/frontend/src/flows/welcome.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  </section>
+  ${footer}`;
+
+export const welcomePage = (props: WelcomeProps): void => {
+  const container = document.getElementById("pageContent") as HTMLElement;
+  render(pageContent(props), container);
+};
+
+export const welcome = (connection: Connection): Promise<LoginData> => {
+  const retryOnError = async (result: LoginFlowResult): Promise<LoginData> => {
```
Comment (peterpeterparker) [inline src/frontend/src/flows/welcome.ts:70]: this is almost the same piece of code as above with `authenticateBox`, maybe it can be refactored (later on) to avoid code duplication?

File: src/frontend/src/flows/welcome.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+      default:
+        unreachable(result);
+        break;
+    }
+  };
+
+  return new Promise<LoginData>((resolve) => {
+    welcomePage({
+      register: async () => {
+        const loginData = await registerIfAllowed(connection).then(
```
Comment (peterpeterparker) [inline src/frontend/src/flows/welcome.ts:93]: can this `.then()` be spared or solved differently?

Reviewed some PR but still hard to follow with the mix of async/await and promises `.then`

File: src/frontend/src/components/authenticateBox.ts (hunk 3 above)
Comment (nmattia) [reply src/frontend/src/components/authenticateBox.ts:26]: ah it is! the type is `bigint`

File: src/frontend/src/components/authenticateBox.ts (hunk 4 above)
Comment (nmattia) [reply src/frontend/src/components/authenticateBox.ts:147]: definitely; will clean this up a bit after this is merged now that I have a better idea of the error handling flow

File: src/frontend/src/index.ts [hunk 10]
```diff
@@ -69,14 +71,15 @@ const init = async () => {
     // Open the management page
     case "manage": {
       // Go through the login flow, potentially creating an anchor.
-      const { userNumber, connection: authenticatedConnection } = await login(
-        connection
-      );
+      const { userNumber, connection: authenticatedConnection } = returning()
+        ? await auth(connection)
```
Comment (peterpeterparker) [inline src/frontend/src/index.ts:75]: nth: not sure if following does not generate a short js?

```
const fnc = returning() ? auth : welcome;
const { userNumber, connection: authenticatedConnection } = await fnc();
```

not sure typescript would be happy with though...

File: src/frontend/src/flows/welcome.ts (hunk 8 above)
Comment (nmattia) [reply src/frontend/src/flows/welcome.ts:70]: Yes, definitely; want to pull the error handling out in both and also avoid nested function calls on error (right now we're recursively calling the function which isn't nice)

File: src/frontend/src/flows/welcome.ts (hunk 9 above)
Comment (nmattia) [reply src/frontend/src/flows/welcome.ts:93]: Will try to get rid of it altogether once I clean up the error handling

File: src/frontend/src/index.ts (hunk 10 above)
Comment (nmattia) [reply src/frontend/src/index.ts:75]: I could also pull out the `await` if it's a matter to make it shorter but I thought it was clearer like this?

File: src/frontend/src/components/authenticateBox.ts [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        withRef(chasmToggleRef, (arrow) =>
+          arrow.classList.remove("c-chasm__button--flipped")
+        );
+      }
+    });
+
+  return html`
+    <p class="t-paragraph t-weak"><span id="alternative-origin-chasm-toggle" class="t-action" @click="${chasmToggle}" >${info} <span ${ref(
+    chasmToggleRef
+  )} class="t-link__icon c-chasm__button">${caretDownIcon}</span></span><br/>
```
Comment (peterpeterparker) [inline src/frontend/src/components/authenticateBox.ts:188]: you need the `br`? if the div is `display: block`?

Comment (peterpeterparker) [reply src/frontend/src/components/authenticateBox.ts:188]: (I don't like br, no particular reason)

Comment (nmattia) [reply src/frontend/src/components/authenticateBox.ts:188]: I don't, good call (although this code was just moved around). Removed!

File: src/frontend/src/flows/welcome.ts (hunk 9 above)
Comment (peterpeterparker) [reply src/frontend/src/flows/welcome.ts:91]: Would really appreciate that! Thanks in advance

File: src/frontend/src/index.ts (hunk 10 above)
Comment (peterpeterparker) [reply src/frontend/src/index.ts:75]: Not really to because of the repetition but can definitely live with.

note: I tried and answered my question, indeed the js would be shorter

source

```

const a = async () => console.log('a');
const b = async () => console.log('b');

export const text = async (tmp: boolean) => {
    const fnc = tmp ? a : b;
    const c = await fnc();   
}

    // vs

export const text = async (tmp: boolean) => {
    const c = tmp ? await a() : await b();
}
```

generates

```
var a=async()=>console.log("a"),c=async()=>console.log("b"),s=async o=>{let n=o?await a():await c()};export{s as text};

var c=async()=>console.log("a"),n=async()=>console.log("b"),a=async o=>{let t=await(o?c:n)()};export{a as text};
```

few bytes saved 😂

File: src/frontend/src/flows/login/index.ts [hunk 12]
```diff
@@ -1,205 +1,16 @@
...
-  render(content.template, container);
+import { LoginData } from "./flowResult";
+import {
+  authenticateBox,
+  AuthTemplates,
+} from "../../components/authenticateBox";
+import { html } from "lit-html";
+
+/* Template for the authbox when authenticating to II */
+const tpls: AuthTemplates = {
```
Comment (frederikrothenberger) [inline src/frontend/src/flows/login/index.ts:10]: Super nit: I'd prefer more descriptive variable names.

File: src/frontend/src/flows/register.ts [hunk 13]
```diff
@@ -21,12 +21,12 @@ const pageContent = html`
...
     <hgroup>
       <div class="c-logo">${icLogo}</div>
-      <p class="t-lead">What device are you using?</p>
+      <p class="t-lead t-centered">What device are you using?</p>
     </hgroup>
     <form id="registerForm" class="l-stack">
       <input
         id="registerAlias"
-        placeholder="Device name"
+        placeholder="Chrome on iPhone"
```
Comment (frederikrothenberger) [inline src/frontend/src/flows/register.ts:29]: This is definitely a better placeholder than what we had before. But this browser separation of webauthn devices is an Apple think. So this might confuse users on Windows / Linux where including the browser name in the device name does not make sense and registering the device using different browsers will fail (due to the webauthn device already being regitstered). How about we suggest just `iPhone` for now?

File: src/frontend/src/showcase.ts [hunk 14]
```diff
@@ -113,58 +114,41 @@ const dummyChallenge: Challenge = {
...
-      console.log("Add device:", userNumber);
-    },
-    recover: () => {
-      console.log("Recover");
-    },
-    register: () => {
-      console.log("Register");
-    },
-    ...returning,
+const welcome = () =>
```
Comment (frederikrothenberger) [inline src/frontend/src/showcase.ts:117]: It would be nice if we had a screenshot / showcase page of the "Authenticate to manage your anchor" page as well.

Also, on that particular page, I'm not sure, labeling the button 'Authorize' makes a lot of sense. I would expect it to be 'Authenticate' or 'Sign in'.

**Review by frederikrothenberger — APPROVED** [review]
LGTM, thanks! Just left small nits.

File: src/frontend/src/flows/register.ts (hunk 13 above)
Comment (nmattia) [reply src/frontend/src/flows/register.ts]: Reverted to what we had previously, we'll need to work on this
